File: src/org/cerion/symcalc/expression/function/list/Range.java
package org.cerion.symcalc.expression.function.list;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class Range extends FunctionExpr {

	public Range(Expr... e) {
		super(FunctionType.RANGE,e);
	}
	
	@Override
	protected Expr evaluate() {
		
		if(get(0).isInteger())
		{
			int num = ((IntegerNum)get(0)).intValue();
			
			ListExpr listResult = new ListExpr();
			int i = 1; //default start at 1
			
			//if 2nd parameter range is num1 to num2
			if(size() > 1 && get(1).isInteger()) 
			{
				i = ((IntegerNum)get(0)).intValue();
				num = ((IntegerNum)get(1)).intValue();
			}
			
			//Step by 1 or 3rd parameter
			int iStep = 1;
			if(size() > 2 && get(2).isInteger())
				iStep = ((IntegerNum)get(2)).intValue();
				
			while(i <= num)
			{
				listResult.add(new IntegerNum(i));
				i = i + iStep;
			}
		
			return listResult;
		}

		return this;
	}
}

File: src/expression/function/list/Range.kt
package org.cerion.symcalc.expression.function.list

import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.function.Function
import org.cerion.symcalc.expression.function.FunctionExpr
import org.cerion.symcalc.expression.number.Integer
import org.cerion.symcalc.expression.number.NumberExpr

// FEAT Implemented with just NumberExpr, this function works on numeric values such as Pi and Sqrt[2]
class Range(vararg e: Expr) : FunctionExpr(Function.RANGE, *e) {

    override fun evaluate(): Expr {
        if (size == 1)
            return range(Integer.ONE, get(0) as NumberExpr, Integer.ONE)

        if (size == 2)
            return range(get(0) as NumberExpr, get(1) as NumberExpr, Integer.ONE)

        return range(get(0) as NumberExpr, get(1) as NumberExpr, get(2) as NumberExpr)
    }

    private fun range(min: NumberExpr, max: NumberExpr, step: NumberExpr): ListExpr {
        val list = mutableListOf<Expr>()
        var current = min

        while(current <= max) {
            list.add(current)
            current += step
        }

        return ListExpr(list)
    }

    override fun validate() {
        validateParameterType(0, ExprType.NUMBER)
        if (size > 1)
            validateParameterType(1, ExprType.NUMBER)
        else if (size > 2)
            validateParameterType(1, ExprType.NUMBER)

        validateParameterRage(1, 3)
    }
}
